### PR TITLE
Improve addLastJoin behaviour

### LINKED ISSUE #1737: SQLQuery.toString should not have side effects
Calling ProjectableSQLQuery.toString registers pending join clauses as a side effect. If one happens to call query.toString() between join() and on() then the query breaks as a result: 

    query.from(x);
    query.leftJoin(y) // toString() call e.g. by debugger
    .on(x.foo.equals(y.foo));
    query.leftJoin(z).on(z.foo.equals(y.foo));

Results in 

    from x 
    left join y 
    left join z on x.foo = y.foo and z.foo = y.foo

ToString with side effects is especially disastrous with modern debuggers as they call toString every time execution pauses. This makes debugging query builders very hard and error-prone.

### PR DESCRIPTION
**Pull request by timowest**
Fixes #1737

### CODE REVIEW

File: querydsl-core/src/main/java/com/mysema/query/DefaultQueryMetadata.java [hunk 1]
```diff
@@ -385,8 +386,7 @@ public boolean equals(Object o) {
 
     @Override
     public int hashCode() {
-        addLastJoin();
-        return Objects.hashCode(flags, groupBy, having, joins, modifiers,
+        return Objects.hashCode(flags, groupBy, having, getJoins(), modifiers,
                 orderBy, params, projection, unique, where);
```
Comment (robertandrewbain) [inline querydsl-core/src/main/java/com/mysema/query/DefaultQueryMetadata.java]: I note that `where` is in hashCode() but not `equals(Object o)`. I would have expected to see it in neither or both.

File: querydsl-core/src/test/java/com/mysema/query/DefaultQueryMetadataTest.java [hunk 2]
```diff
@@ -94,6 +95,20 @@ public void GetJoins2() {
     }
 
     @Test
+    public void GetJoins3() {
```
Comment (robertandrewbain) [inline querydsl-core/src/test/java/com/mysema/query/DefaultQueryMetadataTest.java]: Poorly named.

File: querydsl-core/src/main/java/com/mysema/query/DefaultQueryMetadata.java (hunk 1 above)
Comment (timowest) [inline querydsl-core/src/main/java/com/mysema/query/DefaultQueryMetadata.java]: see line 380

File: querydsl-core/src/test/java/com/mysema/query/DefaultQueryMetadataTest.java (hunk 2 above)
Comment (timowest) [inline querydsl-core/src/test/java/com/mysema/query/DefaultQueryMetadataTest.java]: I agree, but this is a backport, in general we do only fixes in Querydsl 3, and not stylistic improvements

**Pull request merged**